panther_hardware_interfaces/include/panther_hardware_interfaces/canopen_controller.hpp

```diff
@@ -0,0 +1,126 @@
+// Copyright 2023 Husarion sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef PANTHER_HARDWARE_INTERFACES_CANOPEN_CONTROLLER_HPP_
+#define PANTHER_HARDWARE_INTERFACES_CANOPEN_CONTROLLER_HPP_
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <cstdint>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <thread>
+
+#include <lely/coapp/fiber_driver.hpp>
+#include <lely/coapp/master.hpp>
+#include <lely/ev/loop.hpp>
+#include <lely/io2/linux/can.hpp>
+#include <lely/io2/posix/poll.hpp>
+#include <lely/io2/sys/io.hpp>
+#include <lely/io2/sys/sigset.hpp>
+#include <lely/io2/sys/timer.hpp>
+
+#include <panther_hardware_interfaces/roboteq_driver.hpp>
+
+namespace panther_hardware_interfaces
+{
+
+struct CANopenSettings
+{
+  std::string can_interface_name;
+
+  std::uint8_t master_can_id;
+  std::uint8_t front_driver_can_id;
+  std::uint8_t rear_driver_can_id;
+
+  std::chrono::milliseconds pdo_feedback_timeout_ms;
+  std::chrono::milliseconds sdo_operation_timeout_ms;
+};
+
+/**
+ * @brief CANopenController takes care of CANopen communication - creates master controller
+ * and two Roboteq drivers (front and rear)
+ */
+class CANopenController
+{
+public:
+  CANopenController(const CANopenSettings & canopen_settings);
+
+  /**
+   * @brief Starts CANopen communication (in a new thread) and waits for boot to finish
+   *
+   * @exception std::runtime_error if boot fails
+   */
+  void Initialize();
+
+  /**
+   * @brief Stops CANopen communication - sends stop signal and waits
+   */
+  void Deinitialize();
+
+  std::shared_ptr<RoboteqDriver> GetFrontDriver() { return front_driver_; }
+  std::shared_ptr<RoboteqDriver> GetRearDriver() { return rear_driver_; }
+
+private:
+  void InitializeCANCommunication();
+
+  /**
+   * @brief When RT kernel is used configures thread to have kCANopenThreadSchedPriority priority
+   */
+  void ConfigureRT();
```

Comment:
The same method is used in the `panther_gpiod` package. Maybe we could move it to the utils or something?